docs/spec/reactors/consensus/proposer-selection.md

```diff
@@ -2,20 +2,228 @@
+
+|Priority Round| -2 | -1 | 0    | 1   | 2   | 3 | 4 | 5 | Comment
+|--------------| ---|--- |------|---  |---  |---|---|---|--------
+| R end        |    | p2 |      |  p1 |     |   |   |   |update VP(p1)
+| 1            |    |    |      |     | p2  |   |   | p1|A(i)+=VP(i)
+|              | p1 |    |      |     | p2  |   |   |   |A(p1)-= P
+
+#### Validator Removal
+Consider a new example with set.
+
+Validator | p1 | p2 | p3 |
+--------- |--- |--- |--- |
+VP        | 1  | 2  | 3  |
+
+Let's assume that the last round was R and the proposer priorities were as shown in first row (R end) with their sum being 0. At this point p2 is removed and round 1 of next height runs. At the end of the round (penultimate row) the sum of priorities is -2 (minus the priority of the removed process). 
+ 
+
+|Priority Round |-3 | -2 | -1 | 0  | 1   | 2   | 3 | 4 | 5 | Comment
+|---------------|-- | ---|--- |--- |---  |---  |---|---|---|--------
+| R end         |p3 |    |    |    | p1  | p2  |   |   |   |remove p2
+| 1             |   |    |    | p3 |     | p1  |   |   |   |A(i)+=VP(i)
+|               |   | p1 |    | p3 |     |     |   |   |   |A(p1)-= P
+| *new step*    |   |    | p1 |    | p3  |     |   |   |   |A(i) -= avg
+
+The procedure could continue without modifications. However, it is possible that after a sufficiently large number of modifications in validator set, the priority values would group towards maximum or minimum allowed values.
+For this reason, the selection procedure adds another step (see last row) that shifts the current priority values left or right such that the average remains 0.
+
+The modified selection algorithm is:
+
+    def singleProposalSelection (vset):
+        for each validator i in vset:
+            A(i) += VP(i)
+        prop = max(A)
+        A(prop) -= P
+    
+    def proposalSelection (vset, times):
+        for in in range(times):
+        singleProposalSelection(vset)
+
+        // normalize - shift with average priority
+        avg = sum(A(i) for i in vset)/len(vset)
+        for each validator i in vset:
+            A(i) -= avg
+
+Note that the shifting operation happens after singleProposalSelection() has run times times.
+
+#### New Validator
+When a new validator is added same problem as the one described for removal appears, the sum of priorities in the new set is not zero. This is now fixed with the shift step introduced above.
+
+One other issue that needs to be addressed is the following. A validator V with low voting power that has just been elected is moved to the end of the queue. If the validator set is large and/ or with other validators with significantly higher power, validator V will have to wait a lot of rounds to be elected. If V could remove and read itself to the validator set, it would make a significant (albeit unfair) "jump" ahead in the queue, decreasing the numbers of rounds it has to wait. 
```

Comment:
```suggestion
One other issue that needs to be addressed is the following. A validator V with low voting power that has just been elected is moved to the end of the queue. If the validator set is large and/ or with other validators with significantly higher power, validator V will have to wait a lot of rounds to be elected. If V could remove and re-add itself to the validator set, it would make a significant (albeit unfair) "jump" ahead in the queue, decreasing the numbers of rounds it has to wait. 
```

Reply:
changed